KIPs/kip-82.md

```diff
@@ -0,0 +1,376 @@
+    spec.Fee = total_fee
+    spec.Burnt = burnt_fee
+    spec.Proposer = proposer
+    spec.Rewards = {header.RewardBase: proposer}
+    return spec
+
+# Calculates the deferred rewards, which are determined at the end of block processing.
+# Used in reward distribution.
+# Returns a RewardSpec.
+def calc_deferred_reward(header, config, staking_info):
+    minted = config.MintingAmount
+    total_fee, reward_fee, burnt_fee = calc_deferred_fee(header, config)
+
+    proposer, stakers, kgf, kir, split_rem = calc_split(header, config, minted, reward_fee)
+    shares, share_rem = calc_shares(config, staking_info, stakers)
+
+    # Remainder from (CN, KGF, KIR) split goes to KGF
+    kgf += split_rem
+    # Remainder from staker shares goes to Proposer
+    proposer += share_rem
+
+    # If KGF or KIR address is not set, proposer gets the portion.
+    if staking_info.KGFAddr is None:
+        proposer += kgf
+        kgf = 0
+    if staking_info.KIRAddr is None:
+        proposer += kir
+        kir = 0
+
+    spec = RewardSpec()
+    spec.Minted = minted
+    spec.Fee = total_fee
+    spec.Burnt = burnt_fee
+    spec.Proposer = proposer
+    spec.Stakers = stakers
+    spec.Kgf = kgf
+    spec.Kir = kir
+
+    spec.Rewards = defaultdict(int)
+    spec.Rewards[header.RewardBase] += proposer
+    if staking_info.KGFAddr is not None:
+        spec.Rewards[staking_info.KGFAddr] += kgf
+    if staking_info.KIRAddr is not None:
+        spec.Rewards[staking_info.KIRAddr] += kir
+    for reward_addr, reward_amount in shares:
+        spec.Rewards[reward_addr] += reward_amount
+
+    return spec
+
+# Returns (total_fee, reward_fee, burnt_fee)
+def calc_deferred_fee(header, config):
+    # If not DeferredTxFee, fees are already added to the proposer during TX execution.
+    # Therefore, there are no fees to distribute here at the end of block processing.
+    if not config.DeferredTxFee:
+        return (0, 0, 0)
+
+    # Start with the total block gas fee
+    if header.Number >= MAGMA_BLOCK_NUMBER:
+        total = header.GasUsed * header.BaseFee
+    else:
+        total = header.GasUsed * config.UnitPrice
+    reward = total
+    burnt = 0
+
+    # Since Magma, burn half of gas
+    if header.number >= MAGMA_BLOCK_NUMBER:
+        half_fee = reward / 2
+        reward -= half_fee
+        burnt += half_fee
+
+    # If KIP-82 is enabled, burn fees up to BasicReward
+    if header.number >= FORK_BLOCK_NUMBER and config.UseKIP82:
+        minted = config.MintingAmount
+        cn_minted = minted * config.CnRatio // config.TotalRatio
+        basic_reward = cn_minted * config.CnMintedBasicRatio // config.CnMintedTotalRatio
+        if reward < basic_reward:
+            burnt_kip82 = reward
+        else:
+            burnt_kip82 = basic_reward
+
+        reward -= burnt_kip82
+        burnt += burnt_kip82
+
+    return (total, reward, burnt)
+
+# Returns (proposer, stakers, kgf, kir, remaining) amounts
+def calc_split(header, config, minted, fee):
```

Comment:
```suggestion
def calc_block_reward(header, config, minted, fee):
```

Reply:
`calc_block_reward` can be misleading because it's not calculating the block reward = (minted + total_fee). 

It's not easy to come up with a specific name. All I have is abstract names including:
- `split_rewards`
- `calc_portions`
- `distribute_reward`
- `allocate_reward`

And I wanted pure functions prefixed with `get` or `calc` to distinguish them from state-changing functions (`MintKLAY`, `DistributeBlockReward`)

That's why I chose `calc_split`. Let me know if you have other ideas.